packages/expo-web-browser/ios/EXWebBrowser/EXWebBrowser.m

```diff
@@ -0,0 +1,196 @@
+
+- (dispatch_queue_t)methodQueue
+{
+  return dispatch_get_main_queue();
+}
+
+EX_EXPORT_METHOD_AS(openAuthSessionAsync,
+                    openAuthSessionAsync:(NSString *)authURL
+                    redirectURL:(NSString *)redirectURL
+                    resolver:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  [self initializeWebBrowserWithResolver:resolve andRejecter:reject];
+
+
+  if (@available(iOS 11, *)) {
+    NSURL *url = [[NSURL alloc] initWithString: authURL];
+    __weak typeof(self) weakSelf = self;
+    void (^completionHandler)(NSURL * _Nullable, NSError *_Nullable) = ^(NSURL* _Nullable callbackURL, NSError* _Nullable error) {
+      __strong typeof(weakSelf) strongSelf = weakSelf;
+      if (strongSelf) {
+        if (!error) {
+          NSString *url = callbackURL.absoluteString;
+          strongSelf->_redirectResolve(@{
+                                         @"type" : @"success",
+                                         @"url" : url,
+                                         });
+        } else {
+          strongSelf->_redirectResolve(@{
+                                         @"type" : @"cancel",
+                                         });
+        }
+        [strongSelf flowDidFinish];
+      }
+    };
+    _authSession = [[SFAuthenticationSession alloc]
+                    initWithURL:url
+                    callbackURLScheme:redirectURL
+                    completionHandler:completionHandler];
+    [_authSession start];
+  } else {
+    resolve(@{
+              @"type" : @"cancel",
+              @"message" : @"openAuthSessionAsync requires iOS 11 or greater"
+              });
+    [self flowDidFinish];
+  }
+}
+
+
+EX_EXPORT_METHOD_AS(openBrowserAsync,
+                    openBrowserAsync:(NSString *)authURL
+                    resolver:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  if (![self initializeWebBrowserWithResolver:resolve andRejecter:reject]) {
+    return;
+  }
+
+  // Safari View Controller to authorize request
+  NSURL *url = [[NSURL alloc] initWithString:authURL];
+  SFSafariViewController *safariVC = [[SFSafariViewController alloc] initWithURL:url entersReaderIfAvailable:NO];
+  safariVC.delegate = self;
+
+  // By setting the modal presentation style to OverFullScreen, we disable the "Swipe to dismiss"
+  // gesture that is causing a bug where sometimes `safariViewControllerDidFinish` is not called.
+  // There are bugs filed already about it on OpenRadar.
+  [safariVC setModalPresentationStyle: UIModalPresentationOverFullScreen];
+
+  // This is a hack to present the SafariViewController modally
+  UINavigationController *safariHackVC = [[UINavigationController alloc] initWithRootViewController:safariVC];
+  [safariHackVC setNavigationBarHidden:true animated:false];
+
+  UIViewController *currentViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
```

Comment:
is this better than currentViewController provided by EXUtils?

Reply:
Yup, displaying on `currentViewController` freezes app.